Spyder terminal blog (#44)
Quansight-Labs/quansight-labs-site: posts/2019/12/Spyder-terminal.md

<!-- TEASER_END -->

## ES6/JSX syntax
In the first place, we were able to update all the old JavaScript files to a ES6/JS syntax. This change simplified the code base and maintenance for the terminal. Also, it allows us to extend the project in an easier way to new functionalities that the xterm.js API offers. In order to compile this code and run it inside Spyder we migrated our deployment to Webpack. Additionally, the tests for the client terminal were updated to support the new syntax.

## Multiple shells per operative system

A new feature added to the terminal is that the user can configure which shell to use in the terminal. Now the terminal is able to start with `bash`, `sh`, `ksh`, `zsh`, `csh`, `pwsh`, `tcsh`, `screen`, `tmux`, `dash` and `rbash` in UNIX systems and `cmd` and `powershell` in Windows. This option is given on the preferences pane and a restart of Spyder is required to apply the changes.

> ![UNIX shell options for starting the terminal](/images/spyder-terminal/shells.png)

Reviewer: This shows the exact same thing as the other preferences screenshot, so I'd recommend just keeping this one and getting rid of the other, since you already have a cool graphic showing the difference between the three.

This is a great feature because it allows the user to determine their shell interpreter among the ones that are installed in their systems. In this way, Spyder Terminal can be configured with any of the existing shells as long as it is available on their machine. Some examples of the configurable options of the interpreter are shown below, with bash and tcsh.

> 

Reviewer 2: These two screenshots show empty terminals, I don't see the configurable options. Wrong images?

Author: The previous screenshot shows the options that are available in the preferences pane. While these two screenshots display the terminal under two of the configurable shells, I'm not sure if they are necessary.

Reviewer 2: Yeah, probably better to remove the screenshots and the sentence that mentions them, I don't think it adds too much.

![Spyder-terminal running on a bash shell](/images/spyder-terminal/bash.png)

![Spyder-terminal running on a tcsh shell](/images/spyder-terminal/tcsh.png)

> ## Theme configuration

Reviewer: ```suggestion
## Dark theme support and new UI options
```

Be clearer and more descriptive

> Another big change in this new version of the terminal is that now all the Spyder themes options are available. In this way, when a user changes the palette that wants to use with Spyder, automatically the terminal will use the same theme including the font size. Some examples of the new supported themes for the terminal are shown below.

Reviewer: ```suggestion
Another big change in the new version is built-in support for both Spyder's light and dark theme. When you change your Spyder theme or display options, the Terminal automatically adapt its UI, colors and fonts accordingly.
```

> ![Spyder-terminal with minimal color theme](/images/spyder-terminal/minimal.png)

Reviewer: ```suggestion
```

Its rather redundant to show both and it results in a double image block that should be avoided whenever practicable.

> 

Reviewer: ```suggestion
```

As above

![Spyder-terminal with Spyder dark color theme](/images/spyder-terminal/spyder-dark.png)

> Building on the look and feel of the plugin, we also added configurable options for the terminal sounds and the cursor style. The first option determines if the terminal uses a bell sound or not. The second one allows selecting one of three cursor styles.

Reviewer: ```suggestion
Building on the look and feel of the plugin, we also added additional configurable options for sounds and cursors, so you can choose whether the terminal uses a bell sound to alert the user, and can select from one of three text cursor styles.
```

@steff456 You need to describe what the bell sound actually alerts the user to (its not clear in the preferences either).

Again, removed substantial redundancy and wordiness.

> ![Preferences pane for changing the terminal style](/images/spyder-terminal/preferences.png)

Reviewer: ```suggestion
```

Duplicates the above preferences shot (and doesn't even have the menu open showing the cursor options) so it shouldn't be included, especially since it results in a double image (which should be avoided whenever practicable).

> 

Reviewer: ```suggestion
```

Following above.

![Cursor options for Spyder-terminal](/images/spyder-terminal/cursor-style.png)

> ## Shortcut configuration

Reviewer: ```suggestion
## New configurable shortcut
```

Better phrasing and descriptiveness

> Lastly, the shortcuts on the terminal are configurable within the keyboard shortcuts in Spyder 4. In the previous version of the terminal, these shortcuts were hardcoded, but now the user is able to determine which combination of keys she wants to use to copy, paste and open a new terminal.

Reviewer: ```suggestion
Lastly, the terminal's shortcuts for `Copy`, `Paste` and `New terminal` are now no longer hardcoded, and are fully configurable within Spyder's preferences.
```

Yet again, cut out a ton of redundancy.

In recent years "they" has become the accepted word in English to denote a person of indeterminate gender.

![Configurable shortcuts for the Terminal inside Spyder 4](/images/spyder-terminal/shortcuts.png)

> ![Open a new terminal with the configured shortcut](/images/spyder-terminal/new-term.gif)

Reviewer: ```suggestion
```

This is a 2 MB GIF that just shows a new terminal popping open and nothing else. I suggest either replacing this with something more engaging, informative or compelling, or removing it entirely. The user has no way of seeing that a shortcut was pressed so it doesn't really illustrate the point either.

Reviewer: Also, yet again it would avoid a double image.

> The new Spyder-terminal version will enhance the user experience while using Spyder 4. We can't wait for users to install and experiment with the new features available. All the presented work was funded by a NumFOCUS small development grant and Quansight. Please don't forget to keep up with our updates and happy coding!

Reviewer: ```suggestion
We're excited for you to try out Spyder-Terminal 0.3.0 and experiment with the new features available; we'd like to thank [NumFOCUS](https://numfocus.org/) and Quansight for providing the funding and support that made all this work possible. Check back soon for more Spyder updates, and until then, happy coding!
```

Better phrasing, idiomaticity and flow. Also, again, remember to link what you talk about!
